arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -407,35 +412,55 @@ public void setBatchSize(int batchSize) {
   }
 
   private static final class PositionVectorReader extends VectorizedArrowReader {
+    private final Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
+    private final BufferAllocator bufferAllocator = ArrowAllocation.rootAllocator();
+    private final boolean setArrowValidityVector;
     private long rowStart;
+    private int batchSize;
+    private FieldVector vec;
     private NullabilityHolder nulls;
 
+    PositionVectorReader(boolean setArrowValidityVector) {
+      this.setArrowValidityVector = setArrowValidityVector;
+    }
+
     @Override
     public VectorHolder read(VectorHolder reuse, int numValsToRead) {
-      Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
-      FieldVector vec = arrowField.createVector(ArrowAllocation.rootAllocator());
-
-      if (reuse != null) {
-        vec.setValueCount(0);
-        nulls.reset();
+      if (reuse == null) {
+        this.vec = newVector();
+        this.nulls = newNullabilityHolder();
       } else {
-        ((BigIntVector) vec).allocateNew(numValsToRead);
-        for (int i = 0; i < numValsToRead; i += 1) {
-          vec.getDataBuffer().setLong(i * Long.BYTES, rowStart + i);
-        }
-        for (int i = 0; i < numValsToRead; i += 1) {
-          BitVectorHelper.setBit(vec.getValidityBuffer(), i);
+        vec.setValueCount(0);
+      }
+
+      ArrowBuf dataBuffer = vec.getDataBuffer();
+      ArrowBuf validityBuffer = vec.getValidityBuffer();
+
+      for (int i = 0; i < numValsToRead; i += 1) {
+        dataBuffer.setLong((long) i * Long.BYTES, rowStart + i);
+        if (setArrowValidityVector) {
+          BitVectorHelper.setBit(validityBuffer, i);
         }
-        nulls = new NullabilityHolder(numValsToRead);
       }
 
       rowStart += numValsToRead;
       vec.setValueCount(numValsToRead);
-      nulls.setNotNulls(0, numValsToRead);
 
       return new VectorHolder.PositionVectorHolder(vec, MetadataColumns.ROW_POSITION.type(), nulls);
     }
 
+    private BigIntVector newVector() {
```

Comment:
Nit: `newVector` and `newNullabilityHolder` can be static by passing in the allocator and batch size. I'd also make `arrowField` static because it is a field for a static column definition, `MetadataColumns.ROW_POSITION`.